src/main/java/core/basesyntax/serviceimpl/ShopServiceImpl.java

```
@@ -0,0 +1,24 @@
+package core.basesyntax.serviceimpl;
+
+import core.basesyntax.common.FruitTransaction;
+import core.basesyntax.service.OperationStrategy;
+import core.basesyntax.service.ShopService;
+import core.basesyntax.strategy.OperationHandler;
+import java.util.List;
+
+public class ShopServiceImpl implements ShopService {
+    private final OperationStrategy operationStrategy;
+
+    public ShopServiceImpl(OperationStrategy operationStrategy) {
+        this.operationStrategy = operationStrategy;
+    }
+
+    @Override
+    public void process(List<FruitTransaction> transactions) {
+        for (FruitTransaction transaction : transactions) {
+            OperationHandler handler = operationStrategy
+                    .getOperationHandler(transaction.getOperation());
```

Inside `process` you call `operationStrategy.getOperationHandler(transaction.getOperation())` without validating the returned handler. If the strategy does not contain a handler for an operation this will result in a null `handler` and then an NPE on `handler.apply(...)`. The checklist expects invalid data (e.g., incorrect strategy) to be handled. Quote from checklist: "Don't ignore exceptions in try-catch blocks ... Let's rethrow a RuntimeException with an **informative** message and exception object." At minimum, check for a null handler and throw a RuntimeException naming the missing operation and preferably including the transaction or line number context.